{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "IPython.notebook.set_autosave_interval(0)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Autosave disabled\n"
     ]
    }
   ],
   "source": [
    "# basic\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "import math\n",
    "from datetime import date\n",
    "\n",
    "# internet\n",
    "import urllib3\n",
    "import certifi\n",
    "import xmlrpc.client\n",
    "\n",
    "# system / unix\n",
    "import sys\n",
    "import os\n",
    "import re\n",
    "import multiprocessing\n",
    "import gzip as gz\n",
    "import subprocess\n",
    "from subprocess import call\n",
    "from subprocess import Popen, PIPE\n",
    "import uuid\n",
    "import hashlib\n",
    "import cProfile\n",
    "import time\n",
    "from numba import jit\n",
    "import shutil\n",
    "\n",
    "# dataframe and serialization\n",
    "import pandas as pd\n",
    "import json\n",
    "import collections\n",
    "from collections import deque,OrderedDict\n",
    "from functools import reduce\n",
    "\n",
    "# NextGen\n",
    "# parsing bam\n",
    "import pysam\n",
    "# liftover\n",
    "from pyliftover import LiftOver\n",
    "# parsing vcf\n",
    "import vcf as pyvcf\n",
    "from vcf.parser import _Info as VcfInfo, field_counts as vcf_field_counts\n",
    "\n",
    "# Graphics\n",
    "import seaborn as sns\n",
    "\n",
    "# Stats\n",
    "import sklearn\n",
    "from sklearn import linear_model\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from scipy.stats.stats import pearsonr\n",
    "\n",
    "#pd.set_option('display.height', 1300)\n",
    "pd.set_option('display.max_rows', 1000)\n",
    "pd.set_option('display.max_columns', 100)\n",
    "pd.set_option('max_colwidth',1000)\n",
    "%matplotlib inline\n",
    "%autosave 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {},
   "outputs": [],
   "source": [
    "H_dict={\n",
    "    ('AA','UU'):-6.82,\n",
    "    ('AU','AU'):-9.38,\n",
    "    ('UA','UA'):-7.69,\n",
    "    ('CU','AG'):-10.48,\n",
    "    ('CA','UG'):-10.44,\n",
    "    ('GU','AC'):-11.40,\n",
    "    ('GA','UC'):-12.44,\n",
    "    ('CG','CG'):-10.64,\n",
    "    ('GG','CC'):-13.39,\n",
    "    ('GC','GC'):-14.88,\n",
    "    'Init': 3.61,\n",
    "    'tAU': 3.72,\n",
    "    'symSelf':0,\n",
    "}\n",
    "S_dict={\n",
    "    ('AA','UU'):-19.0,\n",
    "    ('AU','AU'):-26.7,\n",
    "    ('UA','UA'):-20.5,\n",
    "    ('CU','AG'):-27.1,\n",
    "    ('CA','UG'):-26.9,\n",
    "    ('GU','AC'):-29.5,\n",
    "    ('GA','UC'):-32.5,\n",
    "    ('CG','CG'):-26.7,\n",
    "    ('GG','CC'):-32.7,\n",
    "    ('GC','GC'):-36.9,\n",
    "    'Init': -1.5,\n",
    "    'tAU': 10.5,\n",
    "    'symSelf':-1.4,\n",
    "}\n",
    "G_37_1M_7_dict={\n",
    "    ('AA','UU'):-0.93,\n",
    "    ('AU','AU'):-1.10,\n",
    "    ('UA','UA'):-1.33,\n",
    "    ('CU','AG'):-2.08,\n",
    "    ('CA','UG'):-2.11,\n",
    "    ('GU','AC'):-2.24,\n",
    "    ('GA','UC'):-2.35,\n",
    "    ('CG','CG'):-2.36,\n",
    "    ('GG','CC'):-3.26,\n",
    "    ('GC','GC'):-3.42,\n",
    "    'Init': 4.09,\n",
    "    'tAU': 0.45,\n",
    "    'symSelf': 0.43,\n",
    "}\n",
    "rc_dict={\n",
    "    'AA':'UU',\n",
    "    'AU':'AU',\n",
    "    'UA':'UA',\n",
    "    'CU':'AG',\n",
    "    'CA':'UG',\n",
    "    'GU':'AC',\n",
    "    'GA':'UC',\n",
    "    'CG':'CG',\n",
    "    'GG':'CC',\n",
    "    'GC':'GC',\n",
    "    'UU':'AA',\n",
    "    'AG':'CU',\n",
    "    'UG':'CA',\n",
    "    'AC':'GU',\n",
    "    'UC':'GA',\n",
    "    'CC':'GG'\n",
    "}\n",
    "GU_dict={\n",
    "    ('UU','GG'):-1.10,\n",
    "    ('GU','GU'):-1.10,\n",
    "    ('UG','UG'):-1.10,\n",
    "    ('GG','UU'):-1.10,\n",
    "        }\n",
    "revcomp_dict={\n",
    "    'A':'U',\n",
    "    'C':'G',\n",
    "    'G':'C',\n",
    "    'U':'A',\n",
    "    'T':'A'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this function unfolds the tuples in the dictionary above for convenience in computation\n",
    "def unfoldDict(indict):\n",
    "    outdict={}\n",
    "    for k in indict:\n",
    "        v=indict[k]\n",
    "        if isinstance(k,tuple):\n",
    "            for u in k:\n",
    "                outdict[u]=v\n",
    "        else:\n",
    "            outdict[k]=v\n",
    "    return outdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "#5'->3'\n",
    "s_dict=unfoldDict(S_dict)\n",
    "h_dict=unfoldDict(H_dict)\n",
    "g_dict=unfoldDict(G_37_1M_7_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-7.295294584688707"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Chen and Znosko, 2013\n",
    "def fgc(seq):\n",
    "    n=0\n",
    "    nGC=0\n",
    "    for c in seq:\n",
    "        n+=1\n",
    "        if c=='G' or c=='C':\n",
    "            nGC+=1\n",
    "    fgc=float(nGC)/float(n)\n",
    "    return fgc\n",
    "    \n",
    "def saltCorrectG(dG,seq,**kwargs):\n",
    "    na1=kwargs.get('NA1',1.0)\n",
    "    na2=kwargs.get('NA2',0.122)\n",
    "#    print(na1,na2)\n",
    "    q0=fgc(seq)\n",
    "    q1=np.log(na2/na1)\n",
    "    q2=np.log(na2)**2\n",
    "    q3=np.log(na1)**2\n",
    "#    print(dG,q1,q2,q3)\n",
    "    G=dG + (0.324*q0 -0.468)*q1 + 0.133*(q2-q3)\n",
    "    return G\n",
    "# paper calculation validation\n",
    "saltCorrectG(-8.62,\"AAGUGAUC\",NA1=1.021,NA2=0.122)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "siRNA_df=pd.read_excel('/nfs/home/mcgold/data/Wuhan-Hu-1/Wuhan_specific_21mers_nodimers.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'gencode33_21' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-205-4efd05207c4e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mgencode33_21\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/nfs/home/mcgold/data/Wuhan-Hu-1/gencode.v33.transcripts_21mers_sorted.tsv'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"\\t\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'gencode33_21' is not defined"
     ]
    }
   ],
   "source": [
    "gencode33_21_df=pd.read_csv('/nfs/home/mcgold/data/Wuhan-Hu-1/gencode.v33.transcripts_21mers_sorted.tsv',sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACACTACAGCTGATGCATACA\n",
      "ACTACAGCTGATGCATACATA\n",
      "ATGTATGCATCAGCTGTAGTG\n",
      "CTGTAGCGACTGTATGCAGCA\n",
      "ACTAGTACTGATGTCGTATAC\n",
      "CTAGTACTGATGTCGTATACA\n",
      "CTGTATACGACATCAGTACTA\n",
      "AGCTGTATACACTATGCGAGC\n",
      "ATAGTACTACAGATAGAGACA\n",
      "CATAGTACTACAGATAGAGAC\n",
      "CGACGACGACTACTAGCGTGC\n",
      "CTCTATCTGTAGTACTATGAC\n",
      "CTGCTCGCATAGTGTATACAG\n",
      "GCTGTATACACTATGCGAGCA\n",
      "GTGTCTCTATCTGTAGTACTA\n",
      "TCATAGTACTACAGATAGAGA\n",
      "TCTGCTCGCATAGTGTATACA\n",
      "AGTCACATCTGTGACATCACA\n",
      "ACTCGTGACATAGCATCTACA\n",
      "CTCGTGACATAGCATCTACAG\n",
      "ATACGACATCAGTACTAGTGC\n",
      "ATCGACATAGCGAGTGTATGC\n"
     ]
    }
   ],
   "source": [
    "siRNA_dict={}\n",
    "for i in siRNA_df.index:\n",
    "    seq=siRNA_df.loc[i,'seq']\n",
    "    try:\n",
    "        if len(seq)>=21:\n",
    "            print(seq)\n",
    "            siRNA_dict[seq]=[0,0]\n",
    "    except:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ACACTACAGCTGATGCATACA': [0, 0],\n",
       " 'ACTACAGCTGATGCATACATA': [0, 0],\n",
       " 'ATGTATGCATCAGCTGTAGTG': [0, 0],\n",
       " 'CTGTAGCGACTGTATGCAGCA': [0, 0],\n",
       " 'ACTAGTACTGATGTCGTATAC': [0, 0],\n",
       " 'CTAGTACTGATGTCGTATACA': [0, 0],\n",
       " 'CTGTATACGACATCAGTACTA': [0, 0],\n",
       " 'AGCTGTATACACTATGCGAGC': [0, 0],\n",
       " 'ATAGTACTACAGATAGAGACA': [0, 0],\n",
       " 'CATAGTACTACAGATAGAGAC': [0, 0],\n",
       " 'CGACGACGACTACTAGCGTGC': [0, 0],\n",
       " 'CTCTATCTGTAGTACTATGAC': [0, 0],\n",
       " 'CTGCTCGCATAGTGTATACAG': [0, 0],\n",
       " 'GCTGTATACACTATGCGAGCA': [0, 0],\n",
       " 'GTGTCTCTATCTGTAGTACTA': [0, 0],\n",
       " 'TCATAGTACTACAGATAGAGA': [0, 0],\n",
       " 'TCTGCTCGCATAGTGTATACA': [0, 0],\n",
       " 'AGTCACATCTGTGACATCACA': [0, 0],\n",
       " 'ACTCGTGACATAGCATCTACA': [0, 0],\n",
       " 'CTCGTGACATAGCATCTACAG': [0, 0],\n",
       " 'ATACGACATCAGTACTAGTGC': [0, 0],\n",
       " 'ATCGACATAGCGAGTGTATGC': [0, 0]}"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "siRNA_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-0.93\n",
      "C C\n"
     ]
    }
   ],
   "source": [
    "map(lambda x: )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "def NN_0(Q):\n",
    "    pos=0\n",
    "    Qlen=len(Q)\n",
    "    Q=Q.replace('T', 'U')\n",
    "    print(Q)\n",
    "# gu mismatches are treated as AU matches\n",
    "#Biochemistry. 1986 Jun 3;25(11):3209-13.\n",
    "#Free energy contributions of G.U and other terminal mismatches to helix stability.\n",
    "#Q:GU S:GU == AU AU\n",
    "    nQ = Qlen -1\n",
    "    dG_QQ=0\n",
    "    if int(nQ/2) < Qlen/2:\n",
    "        nQ+=1\n",
    "    while pos < nQ-1:\n",
    "        nnQ=str(Q[pos])+str(Q[pos+1])\n",
    "        try:\n",
    "            dG_QQ=dG_QQ + g_dict[nnQ]\n",
    "        except:\n",
    "            pass\n",
    "        pos+=1\n",
    "    return(dG_QQ)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [],
   "source": [
    "def NN(Q,S):\n",
    "    #Q and S have to be equal length\n",
    "    pos=0\n",
    "    Qlen=len(Q)\n",
    "    Slen=len(S)\n",
    "    Q=Q.replace('T', 'U')\n",
    "    S=S.replace('T', 'U')\n",
    "# gu mismatches are treated as AU matches\n",
    "#Biochemistry. 1986 Jun 3;25(11):3209-13.\n",
    "#Free energy contributions of G.U and other terminal mismatches to helix stability.\n",
    "#Q:GU S:GU == AU AU\n",
    "    nQ = Qlen -1\n",
    "    dG_QS=0\n",
    "    if int(nQ/2) < Qlen/2:\n",
    "        nQ+=1\n",
    "    nS = Slen -1\n",
    "    if int(nS/2) < Slen/2:\n",
    "        nS+=1\n",
    "    while pos < nQ-1:\n",
    "        nnQ=Q[pos]+str(Q[pos+1])\n",
    "        nnS=str(S[pos])+str(S[pos+1])\n",
    "        pos+=1\n",
    "        try:\n",
    "            rc_nnS=rc_dict[nnS]\n",
    "        except:\n",
    "            rc_nnS='.'\n",
    "            print(nnS)\n",
    "        if rc_nnS==nnQ:\n",
    "            dG_QS=dG_QS+g_dict[nnQ]\n",
    "        try:\n",
    "            if (nnQ,rc_nnS) in GU_dict:\n",
    "                dG_QS=dG_QS+GU_dict[(nnQ,rc_nnS)]\n",
    "        except:\n",
    "            pass\n",
    "    return(dG_QS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GUGUCUCUAUCUGUAGUACUA\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "-39.29"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NN_0('GTGTCTCTATCTGTAGTACTA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rna_rc(seq):\n",
    "    rcSeq=''\n",
    "    for i in seq:\n",
    "        rcSeq=rcSeq + revcomp_dict[i]\n",
    "    return rcSeq[::-1]\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [],
   "source": [
    "siRNA_df=pd.DataFrame.from_dict(siRNA_dict,orient='index')\n",
    "siRNA_df.columns=['dG_virus','dG_host']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACACUACAGCUGAUGCAUACA\n",
      "ACUACAGCUGAUGCAUACAUA\n",
      "AUGUAUGCAUCAGCUGUAGUG\n",
      "CUGUAGCGACUGUAUGCAGCA\n",
      "ACUAGUACUGAUGUCGUAUAC\n",
      "CUAGUACUGAUGUCGUAUACA\n",
      "CUGUAUACGACAUCAGUACUA\n",
      "AGCUGUAUACACUAUGCGAGC\n",
      "AUAGUACUACAGAUAGAGACA\n",
      "CAUAGUACUACAGAUAGAGAC\n",
      "CGACGACGACUACUAGCGUGC\n",
      "CUCUAUCUGUAGUACUAUGAC\n",
      "CUGCUCGCAUAGUGUAUACAG\n",
      "GCUGUAUACACUAUGCGAGCA\n",
      "GUGUCUCUAUCUGUAGUACUA\n",
      "UCAUAGUACUACAGAUAGAGA\n",
      "UCUGCUCGCAUAGUGUAUACA\n",
      "AGUCACAUCUGUGACAUCACA\n",
      "ACUCGUGACAUAGCAUCUACA\n",
      "CUCGUGACAUAGCAUCUACAG\n",
      "AUACGACAUCAGUACUAGUGC\n",
      "AUCGACAUAGCGAGUGUAUGC\n"
     ]
    }
   ],
   "source": [
    "siRNA_df['dG_virus']=siRNA_df.index.map(lambda x: NN_0(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GCAUACACUCGCUAUGUCGAU'"
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_rc('AUCGACAUAGCGAGUGUAUGC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [],
   "source": [
    "siRNA_df['RNAseq_F-OH']=siRNA_df.index.map(lambda x: x.replace('T','U'))\n",
    "siRNA_df['RNAseq_R-OH']=siRNA_df['RNAseq_F-OH'].map(lambda x: rna_rc(x))\n",
    "siRNA_df['RNAseqX_host-OH']='.'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dG_virus</th>\n",
       "      <th>dG_host</th>\n",
       "      <th>RNAseq_F-OH</th>\n",
       "      <th>RNAseq_R-OH</th>\n",
       "      <th>RNAseqX_host-OH</th>\n",
       "      <th>ddG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACACTACAGCTGATGCATACA</th>\n",
       "      <td>-41.91</td>\n",
       "      <td>-27.72</td>\n",
       "      <td>ACACUACAGCUGAUGCAUACA</td>\n",
       "      <td>UGUAUGCAUCAGCUGUAGUGU</td>\n",
       "      <td>AACAGGCAUACGCCACCACAC</td>\n",
       "      <td>14.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTACAGCTGATGCATACATA</th>\n",
       "      <td>-39.99</td>\n",
       "      <td>-26.27</td>\n",
       "      <td>ACUACAGCUGAUGCAUACAUA</td>\n",
       "      <td>UAUGUAUGCAUCAGCUGUAGU</td>\n",
       "      <td>ACCACACAUGGAUGCCACCUA</td>\n",
       "      <td>13.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATGTATGCATCAGCTGTAGTG</th>\n",
       "      <td>-40.77</td>\n",
       "      <td>-28.60</td>\n",
       "      <td>AUGUAUGCAUCAGCUGUAGUG</td>\n",
       "      <td>CACUACAGCUGAUGCAUACAU</td>\n",
       "      <td>ACACCAGCCAUGGCCACCACA</td>\n",
       "      <td>12.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTGTAGCGACTGTATGCAGCA</th>\n",
       "      <td>-44.32</td>\n",
       "      <td>-26.65</td>\n",
       "      <td>CUGUAGCGACUGUAUGCAGCA</td>\n",
       "      <td>UGCUGCAUACAGUCGCUACAG</td>\n",
       "      <td>ACACAGCGGUCACCAGUGGCA</td>\n",
       "      <td>17.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTAGTACTGATGTCGTATAC</th>\n",
       "      <td>-38.48</td>\n",
       "      <td>-22.02</td>\n",
       "      <td>ACUAGUACUGAUGUCGUAUAC</td>\n",
       "      <td>GUAUACGACAUCAGUACUAGU</td>\n",
       "      <td>ACAUGUGGUGAUGUUGUAGAC</td>\n",
       "      <td>16.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTAGTACTGATGTCGTATACA</th>\n",
       "      <td>-38.35</td>\n",
       "      <td>-23.03</td>\n",
       "      <td>CUAGUACUGAUGUCGUAUACA</td>\n",
       "      <td>UGUAUACGACAUCAGUACUAG</td>\n",
       "      <td>AUCACGUCACCACGACUUGUG</td>\n",
       "      <td>15.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTGTATACGACATCAGTACTA</th>\n",
       "      <td>-38.35</td>\n",
       "      <td>-20.98</td>\n",
       "      <td>CUGUAUACGACAUCAGUACUA</td>\n",
       "      <td>UAGUACUGAUGUCGUAUACAG</td>\n",
       "      <td>AAACAGCAUCACCACAUAUGU</td>\n",
       "      <td>17.37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGCTGTATACACTATGCGAGC</th>\n",
       "      <td>-42.53</td>\n",
       "      <td>-24.52</td>\n",
       "      <td>AGCUGUAUACACUAUGCGAGC</td>\n",
       "      <td>GCUCGCAUAGUGUAUACAGCU</td>\n",
       "      <td>AAGAAAAUAACACGUGUGGCC</td>\n",
       "      <td>18.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATAGTACTACAGATAGAGACA</th>\n",
       "      <td>-38.15</td>\n",
       "      <td>-23.41</td>\n",
       "      <td>AUAGUACUACAGAUAGAGACA</td>\n",
       "      <td>UGUCUCUAUCUGUAGUACUAU</td>\n",
       "      <td>AUAACCACGUCUCUCUCUCUG</td>\n",
       "      <td>14.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CATAGTACTACAGATAGAGAC</th>\n",
       "      <td>-38.15</td>\n",
       "      <td>-26.41</td>\n",
       "      <td>CAUAGUACUACAGAUAGAGAC</td>\n",
       "      <td>GUCUCUAUCUGUAGUACUAUG</td>\n",
       "      <td>ACAGAGCGAGACUACGUCUCA</td>\n",
       "      <td>11.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGACGACGACTACTAGCGTGC</th>\n",
       "      <td>-45.54</td>\n",
       "      <td>-21.50</td>\n",
       "      <td>CGACGACGACUACUAGCGUGC</td>\n",
       "      <td>GCACGCUAGUAGUCGUCGUCG</td>\n",
       "      <td>AGGCGCACACCACCACGACCG</td>\n",
       "      <td>24.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTCTATCTGTAGTACTATGAC</th>\n",
       "      <td>-38.15</td>\n",
       "      <td>-25.02</td>\n",
       "      <td>CUCUAUCUGUAGUACUAUGAC</td>\n",
       "      <td>GUCAUAGUACUACAGAUAGAG</td>\n",
       "      <td>ACUGACUGUAGGUGUCCCUCU</td>\n",
       "      <td>13.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTGCTCGCATAGTGTATACAG</th>\n",
       "      <td>-41.22</td>\n",
       "      <td>-26.41</td>\n",
       "      <td>CUGCUCGCAUAGUGUAUACAG</td>\n",
       "      <td>CUGUAUACACUAUGCGAGCAG</td>\n",
       "      <td>ACACCCGUGUCAUGCGCUUGC</td>\n",
       "      <td>14.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTGTATACACTATGCGAGCA</th>\n",
       "      <td>-42.56</td>\n",
       "      <td>-24.96</td>\n",
       "      <td>GCUGUAUACACUAUGCGAGCA</td>\n",
       "      <td>UGCUCGCAUAGUGUAUACAGC</td>\n",
       "      <td>ACUCCGCAUAACACCUAUGGC</td>\n",
       "      <td>17.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGTCTCTATCTGTAGTACTA</th>\n",
       "      <td>-39.29</td>\n",
       "      <td>-28.64</td>\n",
       "      <td>GUGUCUCUAUCUGUAGUACUA</td>\n",
       "      <td>UAGUACUACAGAUAGAGACAC</td>\n",
       "      <td>ACACAGAGGGACACCACUCCC</td>\n",
       "      <td>10.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCATAGTACTACAGATAGAGA</th>\n",
       "      <td>-38.26</td>\n",
       "      <td>-21.96</td>\n",
       "      <td>UCAUAGUACUACAGAUAGAGA</td>\n",
       "      <td>UCUCUAUCUGUAGUACUAUGA</td>\n",
       "      <td>AAGAGGGACACCACAUCACAG</td>\n",
       "      <td>16.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCTGCTCGCATAGTGTATACA</th>\n",
       "      <td>-41.49</td>\n",
       "      <td>-26.41</td>\n",
       "      <td>UCUGCUCGCAUAGUGUAUACA</td>\n",
       "      <td>UGUAUACACUAUGCGAGCAGA</td>\n",
       "      <td>AGCAAGCGCAUGACACGGGUG</td>\n",
       "      <td>15.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGTCACATCTGTGACATCACA</th>\n",
       "      <td>-41.73</td>\n",
       "      <td>-29.32</td>\n",
       "      <td>AGUCACAUCUGUGACAUCACA</td>\n",
       "      <td>UGUGAUGUCACAGAUGUGACU</td>\n",
       "      <td>AACAUCACUGUGGCCACACCA</td>\n",
       "      <td>12.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTCGTGACATAGCATCTACA</th>\n",
       "      <td>-41.33</td>\n",
       "      <td>-26.31</td>\n",
       "      <td>ACUCGUGACAUAGCAUCUACA</td>\n",
       "      <td>UGUAGAUGCUAUGUCACGAGU</td>\n",
       "      <td>AACCUCCAAGACAUUGUUCAC</td>\n",
       "      <td>15.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTCGTGACATAGCATCTACAG</th>\n",
       "      <td>-41.17</td>\n",
       "      <td>-27.38</td>\n",
       "      <td>CUCGUGACAUAGCAUCUACAG</td>\n",
       "      <td>CUGUAGAUGCUAUGUCACGAG</td>\n",
       "      <td>ACACCUCGGCCUCACUGUGCU</td>\n",
       "      <td>13.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATACGACATCAGTACTAGTGC</th>\n",
       "      <td>-40.44</td>\n",
       "      <td>-21.26</td>\n",
       "      <td>AUACGACAUCAGUACUAGUGC</td>\n",
       "      <td>GCACUAGUACUGAUGUCGUAU</td>\n",
       "      <td>AUAGGGUGGAGACGUGGACAA</td>\n",
       "      <td>19.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATCGACATAGCGAGTGTATGC</th>\n",
       "      <td>-41.78</td>\n",
       "      <td>-24.55</td>\n",
       "      <td>AUCGACAUAGCGAGUGUAUGC</td>\n",
       "      <td>GCAUACACUCGCUAUGUCGAU</td>\n",
       "      <td>ACACGUGUAGCUCACACACAG</td>\n",
       "      <td>17.23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       dG_virus  dG_host            RNAseq_F-OH  \\\n",
       "ACACTACAGCTGATGCATACA    -41.91   -27.72  ACACUACAGCUGAUGCAUACA   \n",
       "ACTACAGCTGATGCATACATA    -39.99   -26.27  ACUACAGCUGAUGCAUACAUA   \n",
       "ATGTATGCATCAGCTGTAGTG    -40.77   -28.60  AUGUAUGCAUCAGCUGUAGUG   \n",
       "CTGTAGCGACTGTATGCAGCA    -44.32   -26.65  CUGUAGCGACUGUAUGCAGCA   \n",
       "ACTAGTACTGATGTCGTATAC    -38.48   -22.02  ACUAGUACUGAUGUCGUAUAC   \n",
       "CTAGTACTGATGTCGTATACA    -38.35   -23.03  CUAGUACUGAUGUCGUAUACA   \n",
       "CTGTATACGACATCAGTACTA    -38.35   -20.98  CUGUAUACGACAUCAGUACUA   \n",
       "AGCTGTATACACTATGCGAGC    -42.53   -24.52  AGCUGUAUACACUAUGCGAGC   \n",
       "ATAGTACTACAGATAGAGACA    -38.15   -23.41  AUAGUACUACAGAUAGAGACA   \n",
       "CATAGTACTACAGATAGAGAC    -38.15   -26.41  CAUAGUACUACAGAUAGAGAC   \n",
       "CGACGACGACTACTAGCGTGC    -45.54   -21.50  CGACGACGACUACUAGCGUGC   \n",
       "CTCTATCTGTAGTACTATGAC    -38.15   -25.02  CUCUAUCUGUAGUACUAUGAC   \n",
       "CTGCTCGCATAGTGTATACAG    -41.22   -26.41  CUGCUCGCAUAGUGUAUACAG   \n",
       "GCTGTATACACTATGCGAGCA    -42.56   -24.96  GCUGUAUACACUAUGCGAGCA   \n",
       "GTGTCTCTATCTGTAGTACTA    -39.29   -28.64  GUGUCUCUAUCUGUAGUACUA   \n",
       "TCATAGTACTACAGATAGAGA    -38.26   -21.96  UCAUAGUACUACAGAUAGAGA   \n",
       "TCTGCTCGCATAGTGTATACA    -41.49   -26.41  UCUGCUCGCAUAGUGUAUACA   \n",
       "AGTCACATCTGTGACATCACA    -41.73   -29.32  AGUCACAUCUGUGACAUCACA   \n",
       "ACTCGTGACATAGCATCTACA    -41.33   -26.31  ACUCGUGACAUAGCAUCUACA   \n",
       "CTCGTGACATAGCATCTACAG    -41.17   -27.38  CUCGUGACAUAGCAUCUACAG   \n",
       "ATACGACATCAGTACTAGTGC    -40.44   -21.26  AUACGACAUCAGUACUAGUGC   \n",
       "ATCGACATAGCGAGTGTATGC    -41.78   -24.55  AUCGACAUAGCGAGUGUAUGC   \n",
       "\n",
       "                                 RNAseq_R-OH        RNAseqX_host-OH    ddG  \n",
       "ACACTACAGCTGATGCATACA  UGUAUGCAUCAGCUGUAGUGU  AACAGGCAUACGCCACCACAC  14.19  \n",
       "ACTACAGCTGATGCATACATA  UAUGUAUGCAUCAGCUGUAGU  ACCACACAUGGAUGCCACCUA  13.72  \n",
       "ATGTATGCATCAGCTGTAGTG  CACUACAGCUGAUGCAUACAU  ACACCAGCCAUGGCCACCACA  12.17  \n",
       "CTGTAGCGACTGTATGCAGCA  UGCUGCAUACAGUCGCUACAG  ACACAGCGGUCACCAGUGGCA  17.67  \n",
       "ACTAGTACTGATGTCGTATAC  GUAUACGACAUCAGUACUAGU  ACAUGUGGUGAUGUUGUAGAC  16.46  \n",
       "CTAGTACTGATGTCGTATACA  UGUAUACGACAUCAGUACUAG  AUCACGUCACCACGACUUGUG  15.32  \n",
       "CTGTATACGACATCAGTACTA  UAGUACUGAUGUCGUAUACAG  AAACAGCAUCACCACAUAUGU  17.37  \n",
       "AGCTGTATACACTATGCGAGC  GCUCGCAUAGUGUAUACAGCU  AAGAAAAUAACACGUGUGGCC  18.01  \n",
       "ATAGTACTACAGATAGAGACA  UGUCUCUAUCUGUAGUACUAU  AUAACCACGUCUCUCUCUCUG  14.74  \n",
       "CATAGTACTACAGATAGAGAC  GUCUCUAUCUGUAGUACUAUG  ACAGAGCGAGACUACGUCUCA  11.74  \n",
       "CGACGACGACTACTAGCGTGC  GCACGCUAGUAGUCGUCGUCG  AGGCGCACACCACCACGACCG  24.04  \n",
       "CTCTATCTGTAGTACTATGAC  GUCAUAGUACUACAGAUAGAG  ACUGACUGUAGGUGUCCCUCU  13.13  \n",
       "CTGCTCGCATAGTGTATACAG  CUGUAUACACUAUGCGAGCAG  ACACCCGUGUCAUGCGCUUGC  14.81  \n",
       "GCTGTATACACTATGCGAGCA  UGCUCGCAUAGUGUAUACAGC  ACUCCGCAUAACACCUAUGGC  17.60  \n",
       "GTGTCTCTATCTGTAGTACTA  UAGUACUACAGAUAGAGACAC  ACACAGAGGGACACCACUCCC  10.65  \n",
       "TCATAGTACTACAGATAGAGA  UCUCUAUCUGUAGUACUAUGA  AAGAGGGACACCACAUCACAG  16.30  \n",
       "TCTGCTCGCATAGTGTATACA  UGUAUACACUAUGCGAGCAGA  AGCAAGCGCAUGACACGGGUG  15.08  \n",
       "AGTCACATCTGTGACATCACA  UGUGAUGUCACAGAUGUGACU  AACAUCACUGUGGCCACACCA  12.41  \n",
       "ACTCGTGACATAGCATCTACA  UGUAGAUGCUAUGUCACGAGU  AACCUCCAAGACAUUGUUCAC  15.02  \n",
       "CTCGTGACATAGCATCTACAG  CUGUAGAUGCUAUGUCACGAG  ACACCUCGGCCUCACUGUGCU  13.79  \n",
       "ATACGACATCAGTACTAGTGC  GCACUAGUACUGAUGUCGUAU  AUAGGGUGGAGACGUGGACAA  19.18  \n",
       "ATCGACATAGCGAGTGTATGC  GCAUACACUCGCUAUGUCGAU  ACACGUGUAGCUCACACACAG  17.23  "
      ]
     },
     "execution_count": 234,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "siRNA_df['Tm_virus']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# method one brute force"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "metadata": {},
   "outputs": [],
   "source": [
    "def displace(Q,Qrc):\n",
    "    reX=''\n",
    "    reXrc=''\n",
    "    for bp in Q:\n",
    "        if bp in ['A','C']:\n",
    "            reX=reX+bp\n",
    "        else:\n",
    "            reX=reX+'[ACGTU]'\n",
    "    for bp in Qrc:\n",
    "        if bp in ['A','C']:\n",
    "            reXrc=reXrc + bp\n",
    "        else:\n",
    "            reXrc=bp + '[ACGTU]' + reXrc\n",
    "    return (reX,reXrc)\n",
    "        "
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "hostFile='gencode.v33.transcripts_21mers_sorted.tsv'\n",
    "hostPath=os.path.join('/nfs/home/mcgold/data/Wuhan-Hu-1',hostFile)\n",
    "siRNA_df['dGx_host']=0;\n",
    "for i in siRNA_df.index:\n",
    "    Q=siRNA_df.loc[i,'RNAseq_F-OH']\n",
    "    Qrc=siRNA_df.loc[i,'RNAseq_R-OH']\n",
    "    min_dG_host=0\n",
    "    # replace G and U with any base in the host kmer file for a list of potential off-targets\n",
    "    check_dict=displace(Q)\n",
    "    for i in Q:\n",
    "    f= open(hostPath,'r')\n",
    "    for line in f:\n",
    "        seq=line.split(\"\\t\")[0]\n",
    "        seq=seq.strip()\n",
    "        seq=seq.replace('T', 'U')\n",
    "        rcSeq=rna_rc(seq)\n",
    "\n",
    "        for i in siRNA_df.index:\n",
    "            Q=siRNA_df.loc[i,'RNAseq_F-OH']\n",
    "            Qrc=siRNA_df.loc[i,'RNAseq_R-OH']\n",
    "            min_dG_host=siRNA_df.loc[i,'dG_host']\n",
    "            q1=NN(Q,seq)\n",
    "            q2=NN(Qrc,seq)\n",
    "            q3=NN(Q,rcSeq)\n",
    "            q4=NN(Qrc,seq)\n",
    "            qmin=min(q1,q2,q3,q4)\n",
    "            if qmin < min_dG_host:\n",
    "                siRNA_df.loc[i,'dG_host']=qmin\n",
    "                siRNA_df.loc[i,'RNAseqX_host-OH']=seq\n",
    "    print(siRNA_df)\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking \n",
      "'ACAC[ACGTU]ACA[ACGTU]C[ACGTU][ACGTU]A[ACGTU][ACGTU]CA[ACGTU]ACA|U[ACGTU]G[ACGTU]U[ACGTU]G[ACGTU]U[ACGTU]G[ACGTU]U[ACGTU]G[ACGTU]U[ACGTU]G[ACGTU]U[ACGTU]U[ACGTU]G[ACGTU]U[ACGTU]ACACACA'...\n"
     ]
    }
   ],
   "source": [
    "# method2 differential displacement\n",
    "# 1 select all Host kmers ignoring G mismatches\n",
    "hostFile='gencode.v33.transcripts_21mers_sorted.tsv'\n",
    "hostPath=os.path.join('/nfs/home/mcgold/data/Wuhan-Hu-1',hostFile)\n",
    "siRNA_df['dGx_host']=0;\n",
    "for i in siRNA_df.index:\n",
    "    Q=siRNA_df.loc[i,'RNAseq_F-OH']\n",
    "    Qrc=siRNA_df.loc[i,'RNAseq_R-OH']\n",
    "    min_dG2_host=0\n",
    "    reX,reXrc=displace(Q,Qrc)\n",
    "    sys.stdout.write(\"Checking \\n\\'\" + reX + \"|\" + reXrc + \"\\'...\\n\")\n",
    "# only check displaced oligos\n",
    "    cmd='grep -P \\'' + reX + \"|\" + reXrc + \"\\' \" + hostPath + \" > /tmp/mcgold/X.txt\"\n",
    "    subprocess.call(cmd, shell=True)\n",
    "    cmd='cat /tmp/mcgold/X.txt >> /tmp/mcgold/Xtargets.tsv'\n",
    "    subprocess.call(cmd, shell=True)\n",
    "    for line in f:\n",
    "        seq=line.split(\"\\t\")[0]\n",
    "        seq=seq.strip()\n",
    "        seq=seq.replace('T', 'U')\n",
    "        rcSeq=rna_rc(seq)\n",
    "\n",
    "        for i in siRNA_df.index:\n",
    "            Q=siRNA_df.loc[i,'RNAseq_F-OH']\n",
    "            Qrc=siRNA_df.loc[i,'RNAseq_R-OH']\n",
    "            min_dG_host2=siRNA_df.loc[i,'dG_host']\n",
    "            q1=NN(Q,seq)\n",
    "            q2=NN(Qrc,seq)\n",
    "            q3=NN(Q,rcSeq)\n",
    "            q4=NN(Qrc,seq)\n",
    "            qmin=min(q1,q2,q3,q4)\n",
    "            if qmin < min_dG2_host:\n",
    "                siRNA_df.loc[i,'dGx_host']=qmin\n",
    "                siRNA_df.loc[i,'RNAseqX_host-OH']=seq\n",
    "                sys.stdout.write(\"Xseq found \" + seq + \"\\n\")\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "siRNA_df['ddG']=siRNA_df['dGx_host']-siRNA_df['dG_virus']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "siRNA_df.to_csv('/nfs/home/mcgold/data/Wuhan-Hu-1/siRNA_df.tsv',sep=\"\\t\",header=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  },
  "widgets": {
   "state": {},
   "version": "1.1.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
